compiler/p/codegen/OMRLoadStoreHandler.hpp

```diff
@@ -0,0 +1,319 @@
+     */
+    static void generateStoreNodeSequence(TR::CodeGenerator *cg, TR::Register *srcReg, TR::Node *node, TR::InstOpCode::Mnemonic storeOp, uint32_t length, bool requireIndexForm=false, int64_t extraOffset=0);
+
+    /**
+     * \brief Generate a sequence of instructions to execute a paired store for a given node.
+     *
+     * This method generates a sequence of instructions at the append cursor that perform a 64-bit store using 32-bit
+     * instructions from a register pair at the location specified by the given store node.
+     *
+     * \param cg The code generator
+     * \param srcReg The register pair whose value should be stored
+     * \param node The store node to emit a store for
+     */
+    static void generatePairedStoreNodeSequence(TR::CodeGenerator *cg, TR::Register *srcReg, TR::Node *node);
+
+    /**
+     * \brief Determines whether the provided node represents a "simple" load.
+     *
+     * This method determines whether the provided node represents a "simple" load. Simple loads are those for which it
+     * is correct to emit a single load instruction without any other surrounding instructions or other setup. Such
+     * loads can, as an optimization, be performed within internal control flow with minimal restrictions.
+     *
+     * This method is guaranteed not to emit any instructions or change any global state.
+     *
+     * \param cg The code generator
+     * \param node The node to check
+     *
+     * \return true if the provided node represents a "simple" load; false otherwise
+     */
+    static bool isSimpleLoad(TR::CodeGenerator *cg, TR::Node *node);
+
+    /**
+     * \brief Evaluate the effective address of a "simple" load node into a memory reference
+     *
+     * This method generates a memory reference for a load which was determined to be simple according to isSimpleLoad.
+     * Managing the lifetime of the returned memory reference's registers is left to the caller.
+     *
+     * \warning While the returned memory reference can be used within internal control flow, calling this method may
+     *          emit arbitrary instructions at the append instruction that may not be safe to perform during internal
+     *          control flow. Hence, this method should always be called before beginning internal control flow.
+     *
+     * \param cg The code generator
+     * \param node The node representing a "simple" load which should be evaluated to a memory reference
+     * \param length The length of data represented by the returned memory reference in bytes
+     * \param requireIndexForm true if the returned memory reference must be indexed-form; false otherwise
+     * \param extraOffset An extra offset in bytes to add to the effective address of the store
+     */
+    static NodeMemoryReference generateSimpleLoadMemoryReference(TR::CodeGenerator *cg, TR::Node *node, uint32_t length, bool requireIndexForm=false, int64_t extraOffset=0);
```

Comment:
No requirement for the corresponding simple store APIs?

Reply:
No. This API does not handle all loads, but only accepts "simple" loads as part of certain codegen optimizations that aren't required for correctness, i.e. we have a fallback path for unresolved and other complex cases that the opt doesn't handle. The fundamental difference between a "simple" load and a non-"simple" one is whether or not it is safe to generate in a single instruction. With a load that can be generated in one instruction, evaluators are free to place instructions performing the load inside ICF as an optimization without needing to worry about fixing up dependencies on the ICF end label, e.g. simple read monitors in OpenJ9 repeat the load until the monitor was not taken during the load.